src/lib/Microsoft.Health.Fhir.Ingest.Template/CalculatedFunction/CalculatedFunctionContentTemplateFactory.cs

```diff
@@ -0,0 +1,95 @@
+    {
+        private const string TargetTypeName = "CalculatedContent";
+        private JmesPath _jmesPath;
+
+        public CalculatedFunctionContentTemplateFactory()
+        {
+            /*
+             * TODO Load and register additional custom JmesPath functions. For now, simply create the basic JmesPath object
+             */
+            _jmesPath = new JmesPath();
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Globalization", "CA1303:Do not pass literals as localized parameters", Justification = "Exception message")]
+        public override IContentTemplate Create(TemplateContainer jsonTemplate)
+        {
+            EnsureArg.IsNotNull(jsonTemplate, nameof(jsonTemplate));
+
+            if (!jsonTemplate.MatchTemplateName(TargetTypeName))
+            {
+                return null;
+            }
+
+            if (jsonTemplate.Template?.Type != JTokenType.Object)
+            {
+                throw new InvalidTemplateException($"Expected an object for the template property value for template type {TargetTypeName}.");
+            }
+
+            var calculatedFunctionTemplate = jsonTemplate.Template.ToValidTemplate<CalculatedFunctionContentTemplate>();
+            var measurementExtractor = new MeasurementExtractor(calculatedFunctionTemplate, CreateExpressionEvaluatorFactory(calculatedFunctionTemplate));
+            return measurementExtractor;
+        }
+
+        private IExpressionEvaluatorFactory CreateExpressionEvaluatorFactory(CalculatedFunctionContentTemplate template)
+        {
+            var evaluatorCache = new Dictionary<string, IExpressionEvaluator>();
+            IExpressionEvaluatorFactory expressionEvaluator = new ExpressionEvaluatorFactory(_jmesPath, template.DefaultExpressionLanguage);
+
+            AddExpression(evaluatorCache, template.TypeMatchExpression, expressionEvaluator);
+            AddExpression(evaluatorCache, template.DeviceIdExpression, expressionEvaluator);
+            AddExpression(evaluatorCache, template.PatientIdExpression, expressionEvaluator);
+            AddExpression(evaluatorCache, template.EncounterIdExpression, expressionEvaluator);
+            AddExpression(evaluatorCache, template.TimestampExpression, expressionEvaluator);
+            AddExpression(evaluatorCache, template.CorrelationIdExpression, expressionEvaluator);
+
+            foreach (var valueExpression in template.Values)
+            {
+                AddExpression(evaluatorCache, valueExpression, expressionEvaluator);
+            }
+
+            return new CachingExpressionEvaluatorFactory(new ReadOnlyDictionary<string, IExpressionEvaluator>(evaluatorCache));
+        }
+
+        private void AddExpression(
+            IDictionary<string, IExpressionEvaluator> cache,
+            Expression expression,
+            IExpressionEvaluatorFactory expressionEvaluator)
+        {
+            if (expression != null)
```

Comment:
Thoughts on introducing an ILogger to this class and logging these expressions?  Would be good to have a log for the null condition as well.

Reply:
Will do

Reply:
The `null` case is valid; it may be the user hasn't supplied a value for a non-required expression. What might be useful is to add in a parameter `isRequired` and throw an exception if a required expression is `null`. I do that when performing the actual Normalization operation but doing it earlier when we construct the content template couldn't hurt.